### ISSUE TITLE
Smoketest bams don't withstand workflow with latest hmftools

### ISSUE THREAD

**Issue by alexiswl** [hartwigmedical/gridss-purple-linx#17]
I've created a CWL version of this workflow as can be seen in full [here](https://github.com/alexiswl/gridss-purple-linx/blob/cwl-workflow/cwl/gridss-purple-linx.cwl).

It uses the following tools and versions
* gridss: 2.9.4 
* gripss: 1.8
* amber: 3.5
* cobalt: 1.9
* purple: 2.48
* linx: 1.11

I've tried using the smoke-test bams in this repo to have a small workflow set to test this but they don't seem to even get past the cobalt stage. 

Toil logs error snippet:

```bash
docker \
            run \
            -i \
            --mount=type=bind,source=/efs/toil/workdir/node-d3bec7df-9c13-47ef-88cd-c1235e6df930-70550dd0-abb1-469a-a430-e736b6269ba4/tmp8er6qqbi/3149a482-a499-4435-85d6-91d0fb6ead5b/tbxxrwdqm/tmp-out3hq8au1w,target=/EShPDU \
            --mount=type=bind,source=/efs/toil/workdir/node-d3bec7df-9c13-47ef-88cd-c1235e6df930-70550dd0-abb1-469a-a430-e736b6269ba4/tmp8er6qqbi/3149a482-a499-4435-85d6-91d0fb6ead5b/tka1hgbsg878nt6kg,target=/tmp \
            --mount=type=bind,source=/efs/reference-data/hartwig-nextcloud/GRIDSS-Purple-Linx-Docker/hg19/dbs/gc/GC_profile.1000bp.cnp,target=/var/lib/cwl/stg1835c845-2e20-4d33-b1ef-57d9d4b6b874/GC_profile.1000bp.cnp,readonly \
            --mount=type=bind,source=/efs/input-data/gridss-purple-linx/smoke_test/CPCT12345678R.bam,target=/var/lib/cwl/stg3d1092de-a674-4585-a8bc-58f1544d5f14/CPCT12345678R.bam,readonly \
            --mount=type=bind,source=/efs/input-data/gridss-purple-linx/smoke_test/CPCT12345678R.bam.bai,target=/var/lib/cwl/stg3d1092de-a674-4585-a8bc-58f1544d5f14/CPCT12345678R.bam.bai,readonly \
            --mount=type=bind,source=/efs/input-data/gridss-purple-linx/smoke_test/CPCT12345678T.bam,target=/var/lib/cwl/stg4ca64f94-958c-43b0-925b-b0a9ae6c4038/CPCT12345678T.bam,readonly \
            --mount=type=bind,source=/efs/input-data/gridss-purple-linx/smoke_test/CPCT12345678T.bam.bai,target=/var/lib/cwl/stg4ca64f94-958c-43b0-925b-b0a9ae6c4038/CPCT12345678T.bam.bai,readonly \
            --workdir=/EShPDU \
            --read-only=true \
            --net=host \
            --user=1000:1000 \
            --rm \
            --env=TMPDIR=/tmp \
            --env=HOME=/EShPDU \
            --cidfile=/efs/toil/workdir/node-d3bec7df-9c13-47ef-88cd-c1235e6df930-70550dd0-abb1-469a-a430-e736b6269ba4/tmp8er6qqbi/3149a482-a499-4435-85d6-91d0fb6ead5b/tka1hgbsg6jc0n91_/20201008184700-040931.cid \
            quay.io/biocontainers/hmftools-cobalt:1.9--0 \
            /bin/sh \
            -c \
            'COBALT' '-Xms2000m' '-Xmx14000m' '-gc_profile' '/var/lib/cwl/stg1835c845-2e20-4d33-b1ef-57d9d4b6b874/GC_profile.1000bp.cnp' '-output_dir' 'CPCT12345678_cobalt' '-reference' 'CPCT12345678R' '-reference_bam' '/var/lib/cwl/stg3d1092de-a674-4585-a8bc-58f1544d5f14/CPCT12345678R.bam' '-threads' '4' '-tumor' 'CPCT12345678T' '-tumor_bam' '/var/lib/cwl/stg4ca64f94-958c-43b0-925b-b0a9ae6c4038/CPCT12345678T.bam' '-validation_stringency' 'STRICT'
 /usr/local/bin/COBALT: line 6: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8): No such file or directory
        07:47:01 - COBALT version: 1.9
        07:47:01 - Thread Count: 4, Window Size: 1000, Min Quality 10
        07:47:01 - Reading GC Profile
        07:47:03 - Calculating Read Count from /var/lib/cwl/stg4ca64f94-958c-43b0-925b-b0a9ae6c4038/CPCT12345678T.bam
        07:47:03 - Generating windows on chromosome 1
        07:47:03 - Generating windows on chromosome 2
        07:47:03 - Generating windows on chromosome 3
        07:47:03 - Generating windows on chromosome 4
        07:47:03 - Calculating Read Count from /var/lib/cwl/stg3d1092de-a674-4585-a8bc-58f1544d5f14/CPCT12345678R.bam
        07:47:03 - Generating windows on chromosome 5
        07:47:03 - Generating windows on chromosome 6
        07:47:03 - Generating windows on chromosome 7
        07:47:03 - Generating windows on chromosome 8
        07:47:03 - Generating windows on chromosome 9
        07:47:03 - Generating windows on chromosome 10
        07:47:03 - Generating windows on chromosome 11
        07:47:03 - Generating windows on chromosome 12
        07:47:03 - Generating windows on chromosome 13
        07:47:03 - Generating windows on chromosome 14
        07:47:03 - Generating windows on chromosome 15
        07:47:03 - Generating windows on chromosome 16
        07:47:03 - Generating windows on chromosome 17
        07:47:03 - Generating windows on chromosome 18
        07:47:03 - Generating windows on chromosome 19
        07:47:03 - Generating windows on chromosome 20
        07:47:03 - Generating windows on chromosome 21
        07:47:03 - Generating windows on chromosome 22
        07:47:03 - Generating windows on chromosome X
        07:47:03 - Generating windows on chromosome Y
        07:47:03 - Generating windows on chromosome 1
        07:47:03 - Generating windows on chromosome 2
        07:47:03 - Generating windows on chromosome 3
        07:47:03 - Generating windows on chromosome 4
        07:47:03 - Generating windows on chromosome 5
        07:47:03 - Generating windows on chromosome 6
        07:47:03 - Generating windows on chromosome 7
        07:47:03 - Generating windows on chromosome 8
        07:47:03 - Generating windows on chromosome 9
        07:47:03 - Generating windows on chromosome 10
        07:47:03 - Generating windows on chromosome 11
        07:47:03 - Generating windows on chromosome 12
        07:47:03 - Generating windows on chromosome 13
        07:47:03 - Generating windows on chromosome 14
        07:47:03 - Generating windows on chromosome 15
        07:47:03 - Generating windows on chromosome 16
        07:47:03 - Generating windows on chromosome 17
        07:47:03 - Generating windows on chromosome 18
        07:47:03 - Generating windows on chromosome 19
        07:47:03 - Generating windows on chromosome 20
        07:47:03 - Generating windows on chromosome 21
        07:47:03 - Generating windows on chromosome 22
        07:47:03 - Generating windows on chromosome X
        07:47:03 - Generating windows on chromosome Y
        07:47:03 - Read Count Complete
        07:47:03 - Applying ratio gc normalization
        07:47:04 - Applying ratio diploid normalization
        07:47:04 - Persisting gc read count and reference ratio medians to CPCT12345678_cobalt
        07:47:04 - Persisting cobalt ratios to CPCT12345678_cobalt/CPCT12345678T.cobalt.ratio.tsv
        07:47:04 - Executing R script via command: Rscript /tmp/script12967643593562710289.R CPCT12345678_cobalt/CPCT12345678T.cobalt.ratio.tsv referenceGCDiploidRatio CPCT12345678_cobalt/CPCT12345678R.cobalt.ratio.pcf
        07:47:04 - Executing R script via command: Rscript /tmp/script15869554251597909846.R CPCT12345678_cobalt/CPCT12345678T.cobalt.ratio.tsv tumorGCRatio CPCT12345678_cobalt/CPCT12345678T.cobalt.ratio.pcf
        07:47:06 - Error executing R script. Examine error file /tmp/ratioSegmentation.R12379221662744739860.error for details.
        07:47:06 - Complete
        Exception in thread "main" java.util.concurrent.ExecutionException: java.io.IOException: R execution failed. Unable to complete segmentation.
                at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
                at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
                at com.hartwig.hmftools.cobalt.RatioSegmentation.applySegmentation(RatioSegmentation.java:37)
                at com.hartwig.hmftools.cobalt.CountBamLinesApplication.run(CountBamLinesApplication.java:105)
                at com.hartwig.hmftools.cobalt.CountBamLinesApplication.main(CountBamLinesApplication.java:46)
        Caused by: java.io.IOException: R execution failed. Unable to complete segmentation.
                at com.hartwig.hmftools.cobalt.RatioSegmentation.ratioSegmentation(RatioSegmentation.java:48)
                at com.hartwig.hmftools.cobalt.RatioSegmentation.lambda$applySegmentation$0(RatioSegmentation.java:33)
                at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
                at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
                at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
                at java.base/java.lang.Thread.run(Thread.java:834)
        [2020-10-08T18:47:07+1100] [MainThread] [I] [cwltool] [job cobalt-1.9.cwl] Max memory used: 1825MiB
        [2020-10-08T18:47:07+1100] [MainThread] [W] [cwltool] [job cobalt-1.9.cwl] completed permanentFail
```

Looking at error `/tmp/ratioSegmentation.R12379221662744739860.error` I get

```
During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C"
2: Setting LC_COLLATE failed, using "C"
3: Setting LC_TIME failed, using "C"
4: Setting LC_MESSAGES failed, using "C"
5: Setting LC_MONETARY failed, using "C"
6: Setting LC_PAPER failed, using "C"
7: Setting LC_MEASUREMENT failed, using "C"
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:parallel':

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs

The following objects are masked from 'package:base':

    Filter, Find, Map, Position, Reduce, anyDuplicated, append,
    as.data.frame, basename, cbind, colnames, dirname, do.call,
    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,
    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,
    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,
    tapply, union, unique, unsplit, which, which.max, which.min

Warning: pcf is not run for sample 1 on chromosome arm  because all observations are missing. NA is returned.
Error in data.frame(rep(sampleid[i], nSeg), seg.chrom, seg.arm, pos.start,  :
  arguments imply differing number of rows: 1, 0
Calls: pcf -> data.frame
Execution halted
```

The inputs to the CWL workflow are as shown below:

```json
{
  "sample_name": "CPCT12345678",
  "normal_sample": "CPCT12345678R",
  "tumor_sample": "CPCT12345678T",
  "tumor_bam": {
    "class": "File",
    "location": "/efs/input-data/gridss-purple-linx/smoke_test/CPCT12345678T.bam"
  },
  "normal_bam": {
    "class": "File",
    "location": "/efs/input-data/gridss-purple-linx/smoke_test/CPCT12345678R.bam"
  },
  "snvvcf": {
    "class": "File",
    "location": "/efs/input-data/gridss-purple-linx/smoke_test/CPCT12345678T.somatic_caller_post_processed.vcf.gz"
  },
  "fasta_reference": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/GRIDSS-Purple-Linx-Docker/hg19/refgenomes/Homo_sapiens.GRCh37.GATK.illumina/Homo_sapiens.GRCh37.GATK.illumina.fasta"
  },
  "reference_dict": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/GRIDSS-Purple-Linx-Docker/hg19/refgenomes/Homo_sapiens.GRCh37.GATK.illumina/Homo_sapiens.GRCh37.GATK.illumina.dict"
  },
  "gc_profile": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/GRIDSS-Purple-Linx-Docker/hg19/dbs/gc/GC_profile.1000bp.cnp"
  },
  "blacklist_gridss": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/GRIDSS-Purple-Linx-Docker/hg19/dbs/gridss/ENCFF001TDO.bed"
  },
  "breakend_pon": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/GRIDSS-Purple-Linx-Docker/hg19/dbs/gridss/pon3792v1/gridss_pon_single_breakend.bed"
  },
  "breakpoint_pon": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/GRIDSS-Purple-Linx-Docker/hg19/dbs/gridss/pon3792v1/gridss_pon_breakpoint.bedpe"
  },
  "breakpoint_hotspot": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/GRIDSS/KnownFusionPairs.hg19.bedpe"
  },
  "bafsnps_amber": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/GRIDSS-Purple-Linx-Docker/hg19/dbs/germline_het_pon_hg19/GermlineHetPon.hg19.vcf.gz"
  },
  "known_fusion_data_linx": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/Linx/known_fusion_data.hg19.csv"
  },
  "gene_transcripts_dir_linx": {
    "class": "Directory",
    "location": "/efs/reference-data/hartwig-nextcloud/Linx/ensemble_data_cache_hg19"
  },
  "viral_hosts_file_linx": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/Linx/viral_host_ref.csv"
  },
  "replication_origins_file_linx": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/Linx/heli_rep_origins.bed"
  },
  "line_element_file_linx": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/Linx/line_elements.hg19.csv"
  },
  "fragile_site_file_linx": {
    "class": "File",
    "location": "/efs/reference-data/hartwig-nextcloud/Linx/fragile_sites_hmf.hg19.csv"
  }
}
```

All reference files from are from the [hartwig-nextcloud portal](https://nextcloud.hartwigmedicalfoundation.nl/s/LTiKTd8XxBqwaiC)

**Comment by alexiswl** [comment]
After downgrading cobalt to v1.8 I'm able to get to running up to the linx visualiser stage.

However it seems that most of the inputs are blank and some of the files in the linx step aren't named correctly - probably because they're never properly written to, since they're empty.

From the `gripss-hardfilter step` we see:

```bash
unpigz -c ...CPCT12345678.filtered.vcf.gz
```

Yields
```bash
##fileformat=VCFv4.2
##ALT=<ID=INV,Description="Inversion">
##FILTER=<ID=ASSEMBLY_BIAS,Description="Mismatch between number of directly supporting reads and reads supporting via assembly.">
##FILTER=<ID=ASSEMBLY_ONLY,Description="Variant is supported only by assembly evidence.">
##FILTER=<ID=ASSEMBLY_TOO_FEW_READ,Description="Not enough reads contribute to this assembly as specified by 'assembly.minReads'">
##FILTER=<ID=ASSEMBLY_TOO_SHORT,Description="This assembly is shorter than a read length">
##FILTER=<ID=INSUFFICIENT_SUPPORT,Description="Does not reach the required threshold quality for calling as specified by 'variantcalling.minScore'">
##FILTER=<ID=LOW_QUAL,Description="Low quality call as specified by 'variantcalling.lowQuality'">
##FILTER=<ID=NO_ASSEMBLY,Description="No assembly supporting this variant could be found.">
##FILTER=<ID=NO_RP,Description="Variant does not have any direct read pair support.">
##FILTER=<ID=NO_SR,Description="Variant does not have any direct split read support.">
##FILTER=<ID=PASS,Description="Variant passes all filters">
##FILTER=<ID=PON,Description="Found in panel of normals">
...
##contig=<ID=17,length=81195210>
##contig=<ID=18,length=78077248>
##contig=<ID=19,length=59128983>
##contig=<ID=20,length=63025520>
##contig=<ID=21,length=48129895>
##contig=<ID=22,length=51304566>
##contig=<ID=X,length=155270560>
##contig=<ID=Y,length=59373566>
##contig=<ID=MT,length=16569>
##gridssVersion=2.9.4-gridss
##gripssVersion=1.8
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  CPCT12345678R   CPCT12345678T
```

While there are entries in the recovery vcf going into purple, the purple output is null

```bash
##fileformat=VCFv4.2
##ALT=<ID=INV,Description="Inversion">
##FILTER=<ID=ASSEMBLY_BIAS,Description="Mismatch between number of directly supporting reads and reads supporting via assembly.">
##FILTER=<ID=ASSEMBLY_ONLY,Description="Variant is supported only by assembly evidence.">
##FILTER=<ID=ASSEMBLY_TOO_FEW_READ,Description="Not enough reads contribute to this assembly as specified by 'assembly.minReads'">
##FILTER=<ID=ASSEMBLY_TOO_SHORT,Description="This assembly is shorter than a read length">
##FILTER=<ID=INFERRED,Description="Breakend inferred from copy number transition">
##FILTER=<ID=INSUFFICIENT_SUPPORT,Description="Does not reach the required threshold quality for calling as specified by 'variantcalling.minScore'">
##FILTER=<ID=LOW_QUAL,Description="Low quality call as specified by 'variantcalling.lowQuality'">
##FILTER=<ID=NO_ASSEMBLY,Description="No assembly supporting this variant could be found.">
##FILTER=<ID=NO_RP,Description="Variant does not have any direct read pair support.">
##FILTER=<ID=NO_SR,Description="Variant does not have any direct split read support.">
##FILTER=<ID=PASS,Description="Variant passes all filters">
...
##contig=<ID=19,length=59128983>
##contig=<ID=20,length=63025520>
##contig=<ID=21,length=48129895>
##contig=<ID=22,length=51304566>
##contig=<ID=X,length=155270560>
##contig=<ID=Y,length=59373566>
##contig=<ID=MT,length=16569>
##gridssVersion=2.9.4-gridss
##gripssVersion=1.8
##purpleVersion=2.48
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  CPCT12345678R   CPCT12345678T
```

This ultimately leads to the linx output directory being as follows:
```bash
CPCT12345678T.linx.drivers.tsv
.linx.vis_segments.tsv
CPCT12345678T.linx.clusters.tsv
CPCT12345678T.linx.fusion.tsv
.linx.vis_protein_domain.tsv
CPCT12345678T.linx.viral_inserts.tsv
CPCT12345678T.linx.svs.tsv
linx.version
.linx.vis_sv_data.tsv
.linx.vis_copy_number.tsv
.linx.vis_fusion.tsv
CPCT12345678T.linx.breakend.tsv
CPCT12345678T.linx.links.tsv
.linx.vis_gene_exon.tsv
```

Which means the step collecting the vis data in the expected format fails since it doesn't contain the sample-name prefix as expected and we have a failure at the linx visualiser step.

**Comment by ohofmann** [comment]
@d-cameron or maybe @p-priestley - would love your take on this. Almost ready to run the trifecta on a number of interesting samples. Would love to get this over the finishing line.

**Comment by jonbaber** [comment]
I believe this issue should be fixed in [cobalt 1.10](https://github.com/hartwigmedical/hmftools/tree/master/count-bam-lines#version-history-and-download-links)

**Comment by p-priestley** [comment]
The smoke-test is not in any way meant to be a functional test (just a test of whether the pipeline will run to completion) so an empty SV file is possible (I will check our results).

To investigate the Linx issue, could you please provide the contents of the purple directory so we can test.

**Comment by p-priestley** [comment]
I have confirmed that the SV file is empty also in our smoke test results.   The incorrect file names when the SV file was empty was indeed a bug and is fixed in the next release of LINX (v1.12)
